File: CS/WindowsFormsApplication25/BaseObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Xpo;
using DevExpress.Xpo.Metadata;

namespace WindowsFormsApplication25 {

    public class PropertyData {
        private BaseObject owner;
        private XPMemberInfo member;
        public PropertyData(BaseObject owner, XPMemberInfo member) {
            this.owner = owner;
            this.member = member;
        }
        public string Name { get { return member.Name; } }
        public Type Type { get { return member.MemberType; } }
        public object OldValue { get { return owner.GetOldValue(member); } }
        public object NewValue { get { return member.GetValue(owner); } }
    }
    [NonPersistent]
    public class BaseObject : XPBaseObject {
        public BaseObject(Session s)
            : base(s) {
            oldValues = new Dictionary<XPMemberInfo, object>();
            properties = new List<PropertyData>();
            foreach (XPMemberInfo member in this.ClassInfo.PersistentProperties) {
                properties.Add(new PropertyData(this, member));
            }
        }
        private Dictionary<XPMemberInfo, object> oldValues;
        internal object GetOldValue(XPMemberInfo member) {
            object v;
            if (oldValues.TryGetValue(member, out v)) return v;
            return null;
        }
        private void SaveOldValues() {
            oldValues.Clear();
            foreach (XPMemberInfo member in this.ClassInfo.PersistentProperties) {
                oldValues.Add(member, member.GetValue(this));
            }
        }
        protected override void OnLoaded() {
            base.OnLoaded();
            if (Session.IsObjectToSave(this)) return;
            SaveOldValues();
        }
        protected override void OnSaved() {
            base.OnSaved();
            SaveOldValues();
        }
        private List<PropertyData> properties;
        public List<PropertyData> Properties { get { return properties; } }
        public Type Type { get { return this.GetType(); } }
    }
}

File: VB/WindowsFormsApplication25/BaseObject.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.Xpo
Imports DevExpress.Xpo.Metadata

Namespace WindowsFormsApplication25

    Public Class PropertyData

        Private owner As BaseObject

        Private member As XPMemberInfo

        Public Sub New(ByVal owner As BaseObject, ByVal member As XPMemberInfo)
            Me.owner = owner
            Me.member = member
        End Sub

        Public ReadOnly Property Name As String
            Get
                Return member.Name
            End Get
        End Property

        Public ReadOnly Property Type As Type
            Get
                Return member.MemberType
            End Get
        End Property

        Public ReadOnly Property OldValue As Object
            Get
                Return owner.GetOldValue(member)
            End Get
        End Property

        Public ReadOnly Property NewValue As Object
            Get
                Return member.GetValue(owner)
            End Get
        End Property
    End Class

    <NonPersistent>
    Public Class BaseObject
        Inherits XPBaseObject

        Public Sub New(ByVal s As Session)
            MyBase.New(s)
            oldValues = New Dictionary(Of XPMemberInfo, Object)()
            propertiesField = New List(Of PropertyData)()
            For Each member As XPMemberInfo In ClassInfo.PersistentProperties
                propertiesField.Add(New PropertyData(Me, member))
            Next
        End Sub

        Private oldValues As Dictionary(Of XPMemberInfo, Object)

        Friend Function GetOldValue(ByVal member As XPMemberInfo) As Object
            Dim v As Object
            If oldValues.TryGetValue(member, v) Then Return v
            Return Nothing
        End Function

        Private Sub SaveOldValues()
            oldValues.Clear()
            For Each member As XPMemberInfo In ClassInfo.PersistentProperties
                oldValues.Add(member, member.GetValue(Me))
            Next
        End Sub

        Protected Overrides Sub OnLoaded()
            MyBase.OnLoaded()
            If Session.IsObjectToSave(Me) Then Return
            SaveOldValues()
        End Sub

        Protected Overrides Sub OnSaved()
            MyBase.OnSaved()
            SaveOldValues()
        End Sub

        Private propertiesField As List(Of PropertyData)

        Public ReadOnly Property Properties As List(Of PropertyData)
            Get
                Return propertiesField
            End Get
        End Property

        Public ReadOnly Property Type As Type
            Get
                Return [GetType]()
            End Get
        End Property
    End Class
End Namespace
